File: backend/lib/master-key-frame.js
// Encoding of the outbound `master-key` control frame — the derived master
// key on its way back to native's cache. Split out of `index.js` so the
// encoding can be pinned against `parse-init.js`'s inbound validator without
// booting the backend.

import { Buffer } from "node:buffer";

const MASTER_KEY_BYTE_LENGTH = 32;

/**
 * @param {Buffer} masterKey
 * @returns {{ type: "master-key", masterKey: string }}
 */
export function masterKeyFrame(masterKey) {
  if (
    !Buffer.isBuffer(masterKey) ||
    masterKey.length !== MASTER_KEY_BYTE_LENGTH
  ) {
    throw new Error(
      `masterKey must be a ${MASTER_KEY_BYTE_LENGTH}-byte Buffer, got ` +
        `${Buffer.isBuffer(masterKey) ? `${masterKey.length} bytes` : typeof masterKey}`,
    );
  }
  return { type: "master-key", masterKey: masterKey.toString("base64") };
}

File: android/src/main/java/com/comapeo/core/MasterKeyFrame.kt
package com.comapeo.core

import android.util.Base64

/**
 * Validation of the inbound `master-key` frame, kept free of the service so
 * the gate is unit-testable on the JVM.
 */
internal object MasterKeyFrame {

    const val MASTER_KEY_BYTE_LENGTH = 32

    /** Same shape `backend/lib/parse-init.js` accepts for the outbound field. */
    private val STRICT_BASE64 = Regex("^[A-Za-z0-9\\+/]{43}=$")

    /**
     * Returns the 32 raw bytes of [base64], or null when it is not the strict
     * base64 of a 32-byte key. The caller owns zeroing the result.
     *
     * No trailing-bits round-trip check (unlike `parse-init.js` inbound): the
     * sender is our own backend, whose `Buffer.toString("base64")` always
     * emits the standard encoding.
     *
     * @param decoder test seam — `android.util.Base64` is not available on the JVM.
     */
    fun decode(
        base64: String,
        decoder: (String) -> ByteArray = { Base64.decode(it, Base64.NO_WRAP) },
    ): ByteArray? {
        if (!STRICT_BASE64.matches(base64)) return null
        val bytes = try {
            decoder(base64)
        } catch (_: Throwable) {
            return null
        }
        if (bytes.size != MASTER_KEY_BYTE_LENGTH) {
            bytes.fill(0)
            return null
        }
        return bytes
    }
}
